bundles/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/internal/handler/PHCBridgeHandler.java

```diff
@@ -104,215 +110,405 @@ public void initialize() {
+
+                        if (logger.isDebugEnabled()) {
+                            logger.debug("QueueObject found: {}", qo);
+                        }
+
+                        // put recognized message into queue
+                        if (!receiveQueue.contains(qo)) {
+                            receiveQueue.offer(qo);
+                        }
+
+                        // ignore if message not from EM module
+                    } else {
+                        if (logger.isDebugEnabled()) {
+                            logger.debug("Incoming message not from EM module: {} {} {}", module, sizeToggle, command);
+                        }
+                    }
+                    module = -1;
                 }
+            }
+        }
+    }
 
-                if (!repeat.contains(Arrays.toString(message))) { // avoids multiple processing of the same message
-                    processIncomingMessage(message, size);
-                    repeat.add(Arrays.toString(message));
+    /**
+     * process receive queue
+     */
+    private void processReceiveQueue() {
+        while (true) {
+            if (!receiveQueue.isEmpty()) {
+                QueueObject qo = receiveQueue.poll();
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Consume Receive QueueObject: {}", qo);
                 }
+                handleIncomingCommand(qo.getModuleAddress(), qo.getChannel(), (OnOffType) qo.getCommand());
             }
         }
     }
 
-    private void processQueueLoop() {
+    /**
+     * process send queue
+     */
+    private void processSendQueue() {
         while (true) {
-            long limit = System.currentTimeMillis();
-
-            Map<Long, QueueObject> subQueue = queue.subMap(0L, true, limit, true);
-
-            for (Long key : subQueue.keySet()) {
-                QueueObject qo = subQueue.get(key);
-                queue.remove(key);
-                if (qo.getCounter() < SEND_RETRY_COUNT && !qo.getCommand()
-                        .equals(isChannelOutputStateSet(qo.getModuleType(), qo.getModuleAddress(), qo.getChannel()))) {
-                    qo.increaseCounter();
-
-                    switch (qo.getModuleType()) {
-                        case PHCBindingConstants.CHANNELS_AM:
-                            sendAm(qo.getModuleAddress(), qo.getChannel(), qo.getCommand());
-                            break;
-                        case PHCBindingConstants.CHANNELS_EM_LED:
-                            sendEm(qo.getModuleAddress(), qo.getChannel(), qo.getCommand());
-                            break;
-                        case PHCBindingConstants.CHANNELS_JRM:
-                            sendJRM(qo.getModuleAddress(), qo.getChannel(), qo.getCommand(), qo.getUpDownTime());
-                            break;
-                    }
+            @Nullable
+            QueueObject qo = sendQueue.poll();
```

Comment:
It sounds like you actually would like to use a blocking queue, ie. `BlockingQueue` and use the method `take` which blocks until there is something for you to process. To keep your code readable I would split of the actual processing in its own method.